Android/app/src/main/java/app/intra/SettingsFragment.java

```diff
@@ -0,0 +1,94 @@
+package app.intra;
+
+import android.content.pm.ApplicationInfo;
+import android.content.pm.PackageManager;
+import android.os.Build;
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.v14.preference.MultiSelectListPreference;
+import android.support.v7.preference.PreferenceFragmentCompat;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * UI Fragment for exposing the app preferences.
+ */
+
+public class SettingsFragment extends PreferenceFragmentCompat {
+  private List<String> labels;
+  private List<String> packageNames;
+  private MultiSelectListPreference apps = null;
+
+  /**
+   * Custom pair class for labels (human-readable names) and package names (Java-style package
+   * identifiers).  This class's primary purpose is to expose a custom ordering, where entries
+   * are sorted by label (if non-null) before packageName, and Entries with null labels are placed
+   * after those with non-null labels.
+   */
+  private class Entry implements Comparable<Entry> {
+    public String label;  // Human-readable app name (may be null)
+    public String packageName;  // Java-style package name (non-null)
+
+    Entry(@Nullable String label, @NonNull String packageName) {
+      this.label = label;
+      this.packageName = packageName;
+    }
+
+    @Override
+    public int compareTo(@NonNull Entry other) {
+      if (label == null) {
+        if (other.label == null) {
+          return packageName.compareTo(other.packageName);
+        }
+        return -1;
+      } else if (other.label == null) {
+        return 1;
+      } else if (!label.equals(other.label)) {
+        return label.compareTo(other.label);
+      }
+      return packageName.compareTo(other.packageName);
+    }
+  }
+
+  @Override
+  public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
+    // Load the preferences from an XML resource
+    addPreferencesFromResource(R.xml.preferences);
+    apps = (MultiSelectListPreference)findPreference("pref_apps");
+    // App exclusion relies on VpnService.Builder.addDisallowedApplication, which was added in L.
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+      apps.setEntries(labels.toArray(new String[0]));
+      apps.setEntryValues(packageNames.toArray(new String[0]));
+    } else {
+      apps.setEnabled(false);
+      apps.setSummary(R.string.old_android);
+    }
+  }
+
+  /**
+   * Update this SettingsFragment's list of installed apps for the "Excluded apps" setting.  This
+   * method must be called before the fragment is displayed.
+   * @param pm The global PackageManager.
+   */
+  public void updateInstalledApps(PackageManager pm) {
```

Comment:
Do you think all installed packages is the most sensible?

I've been debating whether some combo of [LauncherApps](https://developer.android.com/reference/android/content/pm/LauncherApps) and system packages might be a better UX, but I think we wouldn't hit our primary use case (download manager)

Reply:
Alas, the download manager runs as the app that is using it, so AFAIK there's no way to blacklist it universally.  I did try!

This code uses [`getInstalledApplications`](https://developer.android.com/reference/android/content/pm/PackageManager#getInstalledApplications(int)), not `getInstalledPackages`, so it only collects the "application packages", not "all installed packages".  However, it does seem to be pulling in a lot of packages that aren't normally user-visible.  I added a check to exclude non-launchable packages, which should make the list shorter and more comprehensible.